### ISSUE TITLE
unnecessary_unwrap not produced in 1.76

### ISSUE THREAD

**Issue by Miha-Rozina** [rust-lang/rust-clippy#12295]
### Summary

The following code doesn't produce a warning for the `unnecessary_unwrap` lint. It works as intended in Rust 1.73

### Lint Name

unnecessary_unwrap

### Reproducer

I tried this code:

```rust
fn foo(option: Option<String>) {
    if option.is_some() {
        println!("{}", option.unwrap());
    }
}
```

I expected to see this happen:
To get the unnecessary_unwrap lint's warning. 

Instead, this happened:
No warnings emitted.

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: aarch64-apple-darwin
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by y21** [comment]
It looks like the lint avoids visiting any nested expressions as soon as it finds code from an external macro (since it `return;`s and doesn't call `walk_expr`), which means that it won't see that `option.unwrap()`, since it is somewhere within the expansion of `println!`:

https://github.com/rust-lang/rust-clippy/blob/237fbdd4da1fee6842cf5c25ab8078acb913696d/clippy_lints/src/unwrap.rs#L302-L306

... but that logic was added 4 years ago, so "It works as intended in Rust 1.73" is surprising to me.

We might just be able to add another call to `walk_expr` in that `if` 🤔 

FWIW, I did make changes to this lint, and the date lines up with 1.73 where the regression could have started, but this particular reproducer doesn't look at all related to my changes. Was this reproducer broken down or slightly modified in a way that hides a different regression?
